src/providers/db2/qgsdb2provider.cpp

```diff
@@ -0,0 +1,1177 @@
+    mSRId = anUri.srid().toInt();
+  else
+    mSRId = -1;
+
+  mWkbType = QGis::fromNewWkbType( anUri.newWkbType() );
+  mValid = true;
+  mSkipFailures = false;
+  int dim; // Not used
+  db2WkbTypeAndDimension( mWkbType, mGeometryColType, dim ); // Get DB2 geometry type name
+
+  mFidColName = anUri.keyColumn();
+  QgsDebugMsg( "mFidColName " + mFidColName );
+  mExtents = anUri.param( "extents" );
+  QgsDebugMsg( "mExtents " + mExtents );
+
+  mUseEstimatedMetadata = anUri.useEstimatedMetadata();
+  QgsDebugMsg( QString( "mUseEstimatedMetadata: '%1'" ).arg( mUseEstimatedMetadata ) );
+  mSqlWhereClause = anUri.sql();
+  QString errMsg;
+  mDatabase = GetDatabase( uri, errMsg );
+  mConnInfo = anUri.connectionInfo();
+
+  if ( !errMsg.isEmpty() )
+  {
+    setLastError( errMsg );
+    QgsDebugMsg( mLastError );
+    mValid = false;
+    return;
+  }
+
+  // Create a query for default connection
+  mQuery = QSqlQuery( mDatabase );
+
+  mSchemaName = anUri.schema();
+
+  mTableName = anUri.table();
+  QStringList sl = mTableName.split( '.' );
+  if ( sl.length() == 2 )  // Never seems to be the case
+  {
+    mSchemaName = sl[0];
+    mTableName = sl[1];
+  }
+
+  QgsDebugMsg( "mSchemaName: '" + mSchemaName + "; mTableName: '" + mTableName );
+
+  if ( !anUri.geometryColumn().isEmpty() )
+    mGeometryColName = anUri.geometryColumn();
+
+  loadFields();
+  UpdateStatistics();
+
+  if ( mGeometryColName.isEmpty() )
+  {
+    // table contains no geometries
+    mWkbType = QGis::WKBNoGeometry;
+    mSRId = 0;
+  }
+
+  //fill type names into sets
+  mNativeTypes
+  // integer types
+  << QgsVectorDataProvider::NativeType( tr( "8 Bytes integer" ), "bigint", QVariant::Int )
+  << QgsVectorDataProvider::NativeType( tr( "4 Bytes integer" ), "integer", QVariant::Int )
+  << QgsVectorDataProvider::NativeType( tr( "2 Bytes integer" ), "smallint", QVariant::Int )
+  << QgsVectorDataProvider::NativeType( tr( "Decimal number (numeric)" ), "numeric", QVariant::Double, 1, 31, 0, 31 )
+  << QgsVectorDataProvider::NativeType( tr( "Decimal number (decimal)" ), "decimal", QVariant::Double, 1, 31, 0, 31 )
+
+  // floating point
+  << QgsVectorDataProvider::NativeType( tr( "Decimal number (real)" ), "real", QVariant::Double )
+  << QgsVectorDataProvider::NativeType( tr( "Decimal number (double)" ), "double", QVariant::Double )
+
+  // string types
+  << QgsVectorDataProvider::NativeType( tr( "Text, fixed length (char)" ), "char", QVariant::String, 1, 254 )
+  << QgsVectorDataProvider::NativeType( tr( "Text, variable length (varchar)" ), "varchar", QVariant::String, 1, 32704 )
+  << QgsVectorDataProvider::NativeType( tr( "Text, variable length large object (clob)" ), "clob", QVariant::String, 1, 2147483647 )
+  //DBCLOB is for 1073741824 double-byte characters, data length should be the same as CLOB (2147483647)?
+  << QgsVectorDataProvider::NativeType( tr( "Text, variable length large object (dbclob)" ), "dbclob", QVariant::String, 1, 1073741824 )
```

Comment:
Does DB2 support date/time/datetime fields? If so, they should be added here too